### PR TITLE
ENH: Accelerate matrix normal sampling using matmul

### PR DESCRIPTION
**Pull request by siddhantwahal**
#### Reference issue
None

#### What does this implement/fix?
Improves performance and readability of `scipy.stats._multivariate.matrix_normal_gen.rvs`.

#### Additional information
This PR introduces changes so that the tensor contractions necessary to sample from a matrix normal distribution rely on `np.einsum` with the `optimize=True` flag, instead of a preset order. In addition to improving readability, this leads to substantial performance gains. 

For instance, to generate `size` samples from an `n` x `p` matrix normal distribution, the current implementation requires:
 

##### Size = 10
|  `einsum`   | p=10 | p=100  | p=1000 |
|--------|------|--------|--------|
| **n=10**  | 0.0  | 0.001   | 0.003   |
| **n=100** | 0.0  | 0.001   | 0.024|
| **n=1000** | 0.002 | 0.023 | 0.473|

| `master` | p=10   | p=100 | p=1000 |
|----------|--------|-------|--------|
| **n=10**     | 0.0    | 0.0   | 0.005  |
| **n=100**   |  0.001  | 0.009  | 0.107  |
| **n=1000**   | 0.193 | 1.709 | 20.081    |

##### Size = 100
| `einsum` | p=10   | p=100 | p=1000 |
|----------|--------|-------|--------|
| **n=10**     | 0.0    | 0.002 | 0.026 |
| **n=100**   | 0.002  | 0.008  | 0.257|
| **n=1000**   | 0.022 | 0.305  | 5.113  |

| `master` | p=10   | p=100 | p=1000 |
|----------|--------|-------|--------|
| **n=10**     | 0.001    | 0.006 | 0.073 |
| **n=100**   | 0.015 | 0.121  | 1.175|
| **n=1000**   | 1.907 | 18.0 | 199.143  |

(Only the tensor calculations were timed (`np.einsum, np.dot, np.tensordot, np.rollaxis`), and employed randomly generated covariances.)

### CODE REVIEW

**Comment by siddhantwahal** [comment]
Should I add a benchmark for this?

File: scipy/stats/_multivariate.py
Comment (larsoner) [inline scipy/stats/_multivariate.py:1148]: You should not need the `np.newaxis` at the beginning, it should be taken care of by broadcasting.

It might also be worth checking if `np.matmul` would be fast, but in theory the `optimize=True` should do better because by using `matmul` you'd have to choose the order ahead of time, unless you know for example that something is square so that the order doesn't matter.

Comment (siddhantwahal) [reply scipy/stats/_multivariate.py]: It turns out that because of the nature of the problem (multiply `n`x`n` times `size`x`n`x`p` times `p`x`p`), `rowchol @ (std_norm @ colchol.T)`, `(rowchol @ std_norm) @ colchol.T`, and `np.einsum('jp,ipq,kq->ijk',  rowchol, std_norm, colchol, optimize=True)`, all require the same number of FLOPS (`2 * size * n * p (n + p)`).

**Comment by sturlamolden** [comment]
The fastest way to sample a multinormal distribution is the transformation

multinorm = mean + L @ stdnorm

where L is the lower Cholesky factor (n x n) of the covariance matrix, mean is an (n x 1) column vector and stdnorm is an (n x 1) column vector or an (n x m) matrix of standard normal deviates. 

For the fastest multiplication we need to use the BLAS functions *TRMV or *TRMM, depending on the shape of stdnorm.

This requires some Cython or Fortran to be maximally efficient.

As for the @ operator, we cannot use this yet as it requires Python >= 3.5.

**Comment by WarrenWeckesser** [comment]
> As for the @ operator, we cannot use this yet as it requires Python >= 3.5.

FYI: Python 3.6 is now the minimum version supported by SciPy, so using `@` is fine.

**Comment by sturlamolden** [comment]
I see, but it still does not call the best BLAS functions for this purpose.

The strategy with two cholesky factorizations also requires one matrix multiplication more than we need. I think we can push the flop count down to 1/4 of what we currently have, if we use the most efficient transformation and optimal BLAS calls.

**Comment by WarrenWeckesser** [comment]
> I see, but it still does not call the best BLAS functions for this purpose.

I edited my comment to say "is allowed" instead of "is fine".  I didn't mean to endorse the use of `@` in this particular case.

**Comment by sturlamolden** [comment]
```@``` operator is beautiful for the readability though.

**Comment by sturlamolden** [comment]
Another thing, for sampling matrix normal using the beforementioned BLAS functions, one would replace

```out = mean + rowchol @ (std_norm @ colchol.T)```

with 

```out = mean + rowchol @ (colchol @ std_norm.T).T```

**Comment by siddhantwahal** [comment]
Thanks for your feedback @sturlamolden and @WarrenWeckesser .

So, the sampling task can be accomplished like so (assuming `rowchol`, `colchol` and `std_norm` are `F_CONTIGUOUS`, and `std_norm` is such that `std_norm.shape` is `(num_rows, num_cols, size)`):

```python
for i in range(std_norm.shape[-1]):
    dtrmm(1.0, rowchol, std_norm[..., i], 
          lower=True, overwrite_b=True)
    dtrmm(1.0, colchol, std_norm[..., i], 
          lower=True, trans_a=True , side=1, 
          overwrite_b=True)
    
return std_norm
```

Is that what you had in mind? This should reduce FLOPS by 1/2.

A question about the design: why does the multivariate normal distribution use `_PSD` class to compute (SVD-based) square roots of the covariance, while this distribution directly uses `scipy.linalg.cholesky`? Was this a conscious choice?

**Comment by sturlamolden** [comment]
Well, almost. The f2py wrapper dtrmm does a probibitive amount of work for small arrays. Rather move the loop to Cython and use the raw function pointers:
```
from scipy.linalg.cython_blas cimport dtrmm, dtrmv
from scipy.linalg.cython_lapack cimport dpotrf
```
As for the multinormal distribution it can use three different methods:  SVD, Jacobi rotations, or Cholesky factorization (see https://github.com/numpy/numpy/blob/master/numpy/random/_generator.pyx). Cholesky is faster than SVD, but can be unstable for near-singular and rank deficient matrices. For example if a Tikhonov regularization is applied to the covariance matrix, you know in advance that Cholesky will be stable.

**Comment by sturlamolden** [comment]
I do not know why Cholesky was choosen for the matrix-normal distribution.

**Comment by siddhantwahal** [comment]
Thanks for the feedback! I'm sorry for the delay, work got unexpectedly busy. I've figured out what changes are necessary and will push them over the weekend.

**Comment by rgommers** [comment]
@sturlamolden would you mind reviewing this one again?

**Comment by sturlamolden** [comment]
@rgommers I am not happy with the way we call BLAS/LAPACK here, which might be my fault because I suggested it (before realizing it would be a mistake).

Currently our Cython API for BLAS/LAPACK only supports 32-bit BLAS/LAPACK. We have started to include 64-bit BLAS/LAPACK in NumPy and SciPy, and we have way of dealing with it seamlessly in C and Cython. There is also a note that Cython API for LAPACK should not be used in SciPy, rather we should link LAPACK directly, something which will also take care of the 32/64-bit issue. Note that the issue here is more complex than simply 32 vs 64 bit Python, as we might use 32 bit LAPACK with 64 bit Python, and (this is the common case) both 32 and 64 bit LAPACK. Then, a second complicating factor is that different compilers use different ABIs. Thankfully, it is all taken care of in the C header ```npy_cblas.h```. 

This is how we should call BLAS or LAPACK from Cython from within SciPy:

```Cython
cimport numpy as cnp
cnp.import_array()

cdef extern from *:
   """
   #include "npy_cblas.h"

    /* BLAS */
    static inline void 
    cython_daxpy(npy_intp n, double alpha, double *x, npy_intp incx, double *y, npy_intp incy)
    {
        CBLAS_INT n_ = n, incx_ = incx, incy_ = incy;
        BLAS_FUNC(daxpy)(&n_, &alpha, x, &incx_, y, &incy_);
    }

    /* LAPACK */
    static inline npy_intp 
    cython_dpttrs(npy_intp n, npy_intp nrhs, double *d, double *e, double *b, npy_intp ldb)
    {
        CBLAS_INT n_ = n, nrhs_ = nrhs, ldb_ = ldb, info = 0;
        BLAS_FUNC(dpttrs)(&n_, &nrhs_, d, e, b, &ldb_, &info);
        return (npy_intp)info;
    }

    """
    cdef void cython_daxpy(cnp.intp_t n, double alpha, double *x, cnp.intp_t incx, double *y, cnp.intp_t incy)
    cdef cnp.intp_t cython_dpttrs(cnp.intp_t n, cnp.intp_t nrhs, double *d, double *e, double *b, cnp.intp_t ldb)

```

To build and link correctly, see what we have done for qhull in ```scipy/spatial/setup.py```.

**Comment by sturlamolden** [comment]
Well actually, when doing something like 
```
    CBLAS_INT n_ = n, nrhs_ = nrhs, ldb_ = ldb, info = 0;
```
we also need to check for integer overflow and raise an error if we overflow the ```CBLAS_INT```, which we might very well do, if we e.g. use 32-bit LAPACK with 64-bit Python. So, the above example needs error handling to be correct as well.

**Comment by sturlamolden** [comment]
So in all its glory, this would be the way to interface BLAS and LAPACK from Cython from *within* SciPy:

```Cython
cimport numpy as cnp
from scipy.linalg import LinAlgError
from libc.limits cimport INT_MAX

cdef extern from *:
   """
   #include "npy_cblas.h"

    /* BLAS */
    static inline void 
    _cython_daxpy(npy_intp n, double alpha, double *x, npy_intp incx, double *y, npy_intp incy)
    {
        CBLAS_INT n_ = n, incx_ = incx, incy_ = incy;
        BLAS_FUNC(daxpy)(&n_, &alpha, x, &incx_, y, &incy_);
    }

    /* LAPACK */
    static inline CBLAS_INT 
    _cython_dpttrs(npy_intp n, npy_intp nrhs, double *d, double *e, double *b, npy_intp ldb)
    {
        CBLAS_INT n_ = n, nrhs_ = nrhs, ldb_ = ldb, info = 0;
        BLAS_FUNC(dpttrs)(&n_, &nrhs_, d, e, b, &ldb_, &info);
        return info;
    }
    """
    ctypedef int CBLAS_INT
    cdef void _cython_daxpy(cnp.intp_t n, double alpha, double *x, cnp.intp_t incx, double *y, cnp.intp_t incy) nogil
    cdef cnp.intp_t _cython_dpttrs(cnp.intp_t n, cnp.intp_t nrhs, double *d, double *e, double *b, cnp.intp_t ldb) nogil

cdef inline int check_cblas_int(cnp.intp_t n) except -1:
    if sizeof(cnp.intp_t n) > sizeof(CBLAS_INT):
        if n > INT_MAX: raise ValueError("integer overflow")
    return 0

cdef int cython_daxpy(cnp.intp_t n, double alpha, double *x, cnp.intp_t incx, double *y, cnp.intp_t incy) except -1:   
    """ Cython wrapper for BLAS daxpy """
    check_cblas_int(n)
    check_cblas_int(incx)
    check_cblas_int(incy)
    with nogil:
        _cython_daxpy(n, alpha, x, incx, y, incy)
    return 0

cdef int cython_dpttrs(cnp.intp_t n, cnp.intp_t nrhs, double *d, double *e, double *b, cnp.intp_t ldb) except -1:
    """ Cython wrapper for LAPACK dpttrs """
    cdef CBLAS_INT info
    check_cblas_int(n)
    check_cblas_int(nrhs)
    check_cblas_int(ldb)
    with nogil:
        info = _cython_dpttrs(n, nrhs, d, e, b, ldb)
    if info < 0: raise LinAlgError("lapack_info = %d" % (info,))
    return 0
```

Releasing the GIL can be expensive, so experiment with this option and consider only doing it for large data sets.

**Comment by sturlamolden** [comment]
For this PR then, follow this pattern and implement it for ```BLAS_FUNC(dpotrf)``` and ```BLAS_FUNC(dtrmm)```, then link the correct BLAS and LAPACK as we have done for qhull (cf. ```scipy/spatial/setup.py```).

File: THANKS.txt
Comment (rlucas7) [inline THANKS.txt:243]: Can this line be removed from the PR? 
See [thread](https://mail.python.org/pipermail/scipy-dev/2020-August/024331.html) for discussion (changes to this file causes merge conflicts adding effort to contributors and maintainers). 
For now, I've opened a [wiki](https://github.com/scipy/scipy/wiki/*THANKS.txt-additions-modifications-for-1.6.0*)
Once the PR is merged, we add thanks to the wiki. Similar process to what we followed in 1.5.0 release.

**Comment by siddhantwahal** [comment]
Sorry everyone, I haven't had time to work on this for a while, and won't for another couple of weeks. I'll definitely make all the changes requested once I get around to it.

**Comment by mdhaber** [comment]
@siddhantwahal It sounds like the originally-proposed changes (e.g. `einsum` or `matmul` instead of `tensordot`) were already big improvement and relatively simple. Would you like to revert to those so we can merge this, then you can work on the even bigger improvements in Cython when you have more time/interest? 

The force-push erased the original commits - do you have them somewhere, or are they easy for you to recreate?

**Comment by siddhantwahal** [comment]
> The force-push erased the original commits - do you have them somewhere, or are they easy for you to recreate?

I think it was just https://github.com/scipy/scipy/commit/7151ababcb1761f4a0ffe7ab21ca121e29401739. I would like to ensure I can still reproduce the timings in the PR description though.

**Comment by mdhaber** [comment]
Perfect. If it sounds good to you, I'll:
- check for performance improvement. Assuming I can reproduce it, I'll continue.
- merge main into this branch
- add a commit that essentially brings this branch to the same state as 7dc3157
- push the changes back here
- merge

This branch will still have the two commits that are here now, so you can revive those someday if you choose.

How does that sound?

**Comment by mdhaber** [comment]
I can confirm the speedup here, but it doesn't look like the random samples are the same (or close). Are they supposed to be the _same_, or is this just a different (but also valid) formulation? If the latter, this is a bit tougher to review. It would be a no-brainer if the output were identical but the time was improved, but if the output is just also valid, I'll need to understand what's going on a bit better - can you help explain it?

**Comment by siddhantwahal** [comment]
> Perfect. If it sounds good to you, I'll:
> 
>     * check for performance improvement. Assuming I can reproduce it, I'll continue.
> 
>     * merge main into this branch
> 
>     * add a commit that essentially brings this branch to the same state as [7dc3157](https://github.com/scipy/scipy/commit/7dc31570bb4bf8633be95d927ff43dde935d14c4)
> 
>     * push (not force-push) the changes back here
> 
>     * merge
> 
> 
> This branch will still have the two commits that are here now, so you can revive those someday if you choose.
> 
> How does that sound?

Sounds great, thank you!

> I can confirm the speedup here, but it doesn't look like the random samples are the same (or close). Are they supposed to be the same, or is this just a different (but also valid) formulation? If the latter, this is a bit tougher to review. It would be a no-brainer if the output were identical but the time was improved, but if the output is just also valid, I'll need to understand what's going on a bit better - can you help explain it?

Thanks for confirming the speedup. The random variates not being the same is indeed concerning. Allow me to take a deeper look over the next couple of days.

File: scipy/stats/_stats.pyx
Comment (sturlamolden) [inline scipy/stats/_stats.pyx:13]: This import should not be used within SciPy.

File: scipy/stats/_stats.pyx
Comment (sturlamolden) [inline scipy/stats/_stats.pyx:14]: This import should not be used within SciPy.

**Review by sturlamolden — CHANGES_REQUESTED** [review]
We cannot use ```cython_blas``` and ```cython_lapack``` internally in SciPy! 

These imports are for external users only. 

As I have previously explain we must link with BLAS and LAPACK during the build stage, as we do in the rest of SciPy and NumPy.

I vote -1 on this PR until this is fixed. I have previously explained how.

**Comment by ev-br** [comment]
> We cannot use cython_blas and cython_lapack internally in SciPy!

> These imports are for external users only.

> As I have previously explain we must link with BLAS and LAPACK during the build stage, as we do in the rest of SciPy and NumPy.

There are multiple internal uses in scipy. Can you give a link to previous explanations?

**Comment by siddhantwahal** [comment]
@sturlamolden @ev-br We're not going forward with the Cython approach, please see https://github.com/scipy/scipy/pull/12312#issuecomment-1217645463

To summarize the state of this PR:
- we need to bring it back to https://github.com/scipy/scipy/commit/7dc31570bb4bf8633be95d927ff43dde935d14c4
- check for correctness and ensure random variates don't change

**Comment by sturlamolden** [comment]
> > We cannot use cython_blas and cython_lapack internally in SciPy!
> 
> > These imports are for external users only.
> 
> > As I have previously explain we must link with BLAS and LAPACK during the build stage, as we do in the rest of SciPy and NumPy.
> 
> There are multiple internal uses in scipy. Can you give a link to previous explanations?

I can try to find a link.

The short explanations is that the Cython wrappers are autogenerated at build time, and thus might not exit yet when the internal Cython module using them is built. It all depends on the order by which the build system operates. We therefore link BLAS and LAPACK directly because using the Cython wrappers can be fragile depending on the build system. When SciPy is built, however, using them will be stable.

**Comment by ev-br** [comment]
A link would indeed be helpful. Possibly in a dedicated issue to not hijack this one, and bring build system wizards into the discussion. 

Your point about build order is taken, so maybe it's a matter of making sure linalg is built first. Or maybe it's trickier and we need to roll back our internal usage of cython_lapack/blas. There definitely is usage, and I'm not aware of any real world problems with either old np.distutils or new meson based build systems. But again, I'm not claiming expertise in this, so best engage build experts.

**Comment by mdhaber** [comment]
Would you like me to merge in main / 7dc3157 and push back here, @siddhantwahal?

**Comment by siddhantwahal** [comment]
> Would you like me to merge in main / [7dc3157](https://github.com/scipy/scipy/commit/7dc31570bb4bf8633be95d927ff43dde935d14c4) and push back here, @siddhantwahal?

Yes, please go ahead.

**Comment by siddhantwahal** [comment]
> I can confirm the speedup here, but it doesn't look like the random samples are the same (or close). Are they supposed to be the _same_, or is this just a different (but also valid) formulation? If the latter, this is a bit tougher to review. It would be a no-brainer if the output were identical but the time was improved, but if the output is just also valid, I'll need to understand what's going on a bit better - can you help explain it?

@mdhaber The difference was due to the the way the standard normal samples were being created. Essentially, the problem boils down to me not realizing that:
```
[nav] In [3]: np.allclose(
         ...:     np.random.default_rng(0).standard_normal(size=(2, 3)),
         ...:     np.random.default_rng(0).standard_normal(size=(3, 2)).T
         ...: )
         ...:
Out[3]: False
```
The fix is simple. Instead of `std_norm = random_state.standard_normal(size=(size, dims[0], dims[1]))`, we'll just do `random_state.standard_normal(size=(dims[1], size, dims[0])).transpose(1, 2, 0)`.

I'll push a fix after you're done merging in `main`.

**Comment by mdhaber** [comment]
All set @siddhantwahal!

**Comment by siddhantwahal** [comment]
@mdhaber Thanks for updating the PR. I just pushed a few commits that ensure that `matrix_normal.rvs` generates the sample sample stream.

**Comment by mdhaber** [comment]
Thanks @siddhantwahal. The random stream is essentially unchanged, so although I haven't thought through why the calculations are equivalent, it's clear that they are. It's only a few lines that are changes, so it seems safe enough.

For the cases suggested at the top, here is a log-log plot of the relative execution time (between this PR and main) vs the absolute execution time in main. 
image

The more time the calculation takes, the greater the improvement made by this PR, which is 30~40 times faster for the slowest cases. The only time this PR is consistently slower is in the very smallest cases.

File: scipy/stats/tests/test_multivariate.py
Comment (mdhaber) [inline scipy/stats/tests/test_multivariate.py:524]: I prefer to keep PRs focused, but I guess this is fine since it's simple. Two other things that would be improvements:

- matching the error message
- if the array of ones were stored, these could fit on the same line

But let's merge as-is.

File: scipy/stats/tests/test_multivariate.py
Comment (mdhaber) [inline scipy/stats/tests/test_multivariate.py:697]: I confirmed that these are the numbers produced in main.

**Pull request merged**
